pkg/destroy/azure/azure.go

```diff
@@ -92,26 +93,110 @@ func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.
 // Run is the entrypoint to start the uninstall process.
 func (o *ClusterUninstaller) Run() error {
 	o.configureClients()
+
 	group := o.InfraID + "-rg"
-	o.Logger.Debug("deleting public records")
-	if err := deletePublicRecords(context.TODO(), o.zonesClient, o.recordsClient, o.privateZonesClient, o.privateRecordSetsClient, o.Logger, group); err != nil {
-		o.Logger.Debug(err)
-		return errors.Wrap(err, "failed to delete public DNS records")
+	returnChannel := make(chan deleteResponse)
+
+	deletePublicRecordsFunc := func(name string, channel chan deleteResponse) {
+		o.Logger.Debugf("deleteing %s", name)
+		response := deleteResponse{name: name, err: nil}
+		response.err = deletePublicRecords(context.TODO(), o.zonesClient, o.recordsClient, o.privateZonesClient, o.privateRecordSetsClient, o.Logger, group)
+		if response.err != nil {
+			o.Logger.Debug(response.err)
+			response.err = errors.Wrap(response.err, "failed to delete public DNS records")
+		}
+		channel <- response
+	}
+	deleteResourceGroupFunc := func(name string, channel chan deleteResponse) {
+		o.Logger.Debugf("deleteing %s", name)
+		response := deleteResponse{name: name, err: nil}
+		response.err = deleteResourceGroup(context.TODO(), o.resourceGroupsClient, o.Logger, group)
+		if response.err != nil {
+			o.Logger.Debug(response.err)
+			response.err = errors.Wrap(response.err, "failed to delete resource group")
+		}
+		channel <- response
 	}
-	o.Logger.Debug("deleting resource group")
-	if err := deleteResourceGroup(context.TODO(), o.resourceGroupsClient, o.Logger, group); err != nil {
-		o.Logger.Debug(err)
-		return errors.Wrap(err, "failed to delete resource group")
+	deleteApplicationRegistrationsFunc := func(name string, channel chan deleteResponse) {
+		o.Logger.Debugf("deleteing %s", name)
+		response := deleteResponse{name: name, err: nil}
+		response.err = deleteApplicationRegistrations(context.TODO(), o.applicationsClient, o.serviceprincipalsClient, o.Logger, o.InfraID)
+		if response.err != nil {
+			o.Logger.Debug(response.err)
+			response.err = errors.Wrap(response.err, "failed to delete application registrations and their service principals")
+		}
+		channel <- response
+	}
+
+	deleteJobs := []deleteJob{
+		{name: "public records", deleteFunc: deletePublicRecordsFunc},
+		{name: "resource group", deleteFunc: deleteResourceGroupFunc},
+		{name: "application registrations", deleteFunc: deleteApplicationRegistrationsFunc},
 	}
-	o.Logger.Debug("deleting application registrations")
-	if err := deleteApplicationRegistrations(context.TODO(), o.applicationsClient, o.serviceprincipalsClient, o.Logger, o.InfraID); err != nil {
-		o.Logger.Debug(err)
-		return errors.Wrap(err, "failed to delete application registrations and their service principals")
+
+	for _, job := range deleteJobs {
+		go job.deleteFunc(job.name, returnChannel)
+	}
+
+	loop := true
+	timeout := time.Duration(30)
+	minutes := 0 * time.Minute
+
+	for {
```

Comment:
Is there a simplification we can make by using like `"k8s.io/apimachinery/pkg/util/wait"` so that we are not writing this channel based look for retrying 2 functions. Keeping it simple would be my preference.

If this ends up being 
```
ctx := new context for 30 mins

wait.PollImmediateUntil (ctx)
  err <- deletePublicRecords
  if err, log at debug and retry.

wait.PollImmediateUntil (ctx)
  err <- deleteResourceGroup
  if err, log at debug and retry.

wait.PollImmediateUntil (ctx)
  err <- deleteServicePrincipkes
  if err, log at debug and retry.
```

i think that would be a good start..

Reply:
I've done the equivalent of what you've suggested here. Fixed.